### PR TITLE
Cache swift authentication token in memcache

### PR DESCRIPTION
**Pull request by icewind1991**
Prevents having to do an authentication roundtrip every request.

Requires
- [x] https://github.com/nextcloud/server/pull/4112
- [x] https://github.com/nextcloud/3rdparty/pull/45

### CODE REVIEW

**Comment by icewind1991** [comment]
https://github.com/nextcloud/server/commit/22eeff687c25e5cca734a0d313423def14f79610 is relevant part, the rest is #4112

File: lib/private/Files/ObjectStore/Swift.php [hunk 1]
```diff
@@ -76,17 +91,64 @@ protected function init() {
 			return;
 		}
 
-		// the OpenCloud client library will default to 'cloudFiles' if $serviceName is null
-		$serviceName = null;
+		$cachedTokenString = $this->memcache->get('token');
+		if ($cachedTokenString) {
+			$cachedToken = unserialize($cachedTokenString);
```
Comment (LukasReschke) [inline lib/private/Files/ObjectStore/Swift.php]: https://github.com/nextcloud/3rdparty/blob/f5555fef8e80d8380efb44dc8b7622a1de573c15/rackspace/php-opencloud/lib/OpenCloud/OpenStack.php#L398-L409 claims this is an array. Can we use `json_encode` and `json_decode` thus? :)

Comment (icewind1991) [inline lib/private/Files/ObjectStore/Swift.php]: the array contains objects, so we can't json it

**Comment by MorrisJobke** [comment]
@icewind1991 Conflicts :/

**Comment by icewind1991** [comment]
rebased

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
Updated the merged 3rdparty submodule.

**Comment by icewind1991** [comment]
@LukasReschke switched to json

@MorrisJobke please retest

**Review by MorrisJobke — APPROVED** [review]
Still works 👍

**Review by rullzer — APPROVED** [review]

**Pull request merged**